### PR TITLE
[mrp] Fix #15799 - add exponential backoff.

### LINKED ISSUE #15799: [msg] Exponential backoff should be implemented according to specification
#### Problem
The exponential backoff algorithm in the SDK does not match the specification.

 #### Proposed Solution
Fix and align behavior with spec.

### LINKED ISSUE #15800: [msg] [mrp] MRP_ACTIVE_TIMEOUT seems to never be used
#### Problem
The sdk doesn't fully implement MRP retransmission timing calculations.

Per slack discussion: https://app.slack.com/team/U01CPER8YCA
There are tickets like this that show some symptoms of this missing support:
https://github.com/project-chip/connectedhomeip/issues/12680

 #### Proposed Solution
Implement ACTIVE/IDLE behavior with pending spec updates to track peer active state per-session rather than per-exchange.

### PR DESCRIPTION
**Pull request by turon**
#### Problem
Fix #15799 - implement exponential backoff in MRP..

#15800 is related, but would come as a separate PR.

#### Change overview
Implements the backoff algorithm as specified but in fixed point:

```
t = i⋅MRP_BACKOFF_BASE^max(0,n−MRP_BACKOFF_THRESHOLD) ⋅ (1.0+random(0,1)⋅MRP_BACKOFF_JITTER)

Where:

  t = the resultant retransmission timeout for this transmission
  n = the transmission count of the message, starting with 0 
  i = the base retry interval for the Exchange (either IDLE or ACTIVE)

MRP_BACKOFF_BASE      | 1.6  | The base number for the exponential backoff equation.
MRP_BACKOFF_JITTER    | 0.25 | The scaler for random jitter in the backoff equation.
MRP_BACKOFF_THRESHOLD | 1    | # of retransmissions before transition from linear to exponential backoff.
```

#### Testing
How was this tested? (at least one bullet point required)
* CI
* Includes Unit test
* Working on unit tests in draft phase

### CODE REVIEW

File: src/messaging/ReliableMessageMgr.cpp [hunk 1]
```diff
@@ -187,11 +187,47 @@ CHIP_ERROR ReliableMessageMgr::AddToRetransTable(ReliableMessageContext * rc, Re
     return CHIP_NO_ERROR;
 }
 
+System::Clock::Timestamp ReliableMessageMgr::GetBackoff(System::Clock::Timestamp backoffBase, uint8_t sendCount)
+{
+#define MRP_BACKOFF_JITTER_BASE 1000
```
Comment (andy31415) [inline src/messaging/ReliableMessageMgr.cpp:192]: could we use C++-style `constexpr` constants instead?

File: src/messaging/ReliableMessageMgr.h [hunk 2]
```diff
@@ -102,6 +102,16 @@ class ReliableMessageMgr
...
 
+    /**
+     *  Calculate the backoff timer for the retransmission.
+     *
+     *  @param[in]   backoffBase    The base interval to use for the backoff calculation, either the active or idle interval.
+     *  @param[in]   sendCount      Count of how many times this message has been sent, including the current attempt.
+     *
+     *  @retval  The backoff time value, including jitter.
+     */
+    System::Clock::Timestamp GetBackoff(System::Clock::Timestamp backoffBase, uint8_t sendCount);
```
Comment (andy31415) [inline src/messaging/ReliableMessageMgr.h:113]: Does this depend on any member variables? I assume not, in which case we could make this static or not be a part of this class at all and be a free standing function.

File: src/messaging/tests/TestReliableMessageProtocol.cpp [hunk 3]
```diff
@@ -1443,6 +1443,78 @@ void CheckLostStandaloneAck(nlTestSuite * inSuite, void * inContext)
...
+    },
+};
+
+const unsigned theBackoffComplianceTestVectorLength =
+    sizeof(theBackoffComplianceTestVector) / sizeof(struct BackoffComplianceTestVector);
+
+void CheckGetBackoff(nlTestSuite * inSuite, void * inContext)
+{
+    TestContext & ctx       = *reinterpret_cast<TestContext *>(inContext);
+    ReliableMessageMgr * rm = ctx.GetExchangeManager().GetReliableMessageMgr();
```
Comment (andy31415) [inline src/messaging/tests/TestReliableMessageProtocol.cpp:1505]: Please make the method static or a free standing function, so you do not need to access it via a ReliableMessageManager object in tests.

File: src/messaging/ReliableMessageMgr.cpp [hunk 4]
```diff
@@ -187,11 +187,47 @@ CHIP_ERROR ReliableMessageMgr::AddToRetransTable(ReliableMessageContext * rc, Re
...
+#define MRP_BACKOFF_JITTER_MAX 250
+#define MRP_BACKOFF_BASE_NUMERATOR 16
+#define MRP_BACKOFF_BASE_DENOMENATOR 10
+#define MRP_BACKOFF_THRESHOLD 1
+
+    System::Clock::Timestamp backoff = backoffBase;
+
+    // Generate fixed point equivalent of 1.6^retryCount
+    int retryCount = sendCount - MRP_BACKOFF_THRESHOLD;
```
Comment (andy31415) [inline src/messaging/ReliableMessageMgr.cpp:201]: could we use std::clamp here to place it between 0 and 4?

File: src/messaging/tests/TestReliableMessageProtocol.cpp [hunk 5]
```diff
@@ -1443,6 +1443,78 @@ void CheckLostStandaloneAck(nlTestSuite * inSuite, void * inContext)
...
+
+    for (unsigned i = 0; i < theBackoffComplianceTestVectorLength; i++)
+    {
+        struct BackoffComplianceTestVector * test = &theBackoffComplianceTestVector[i];
+        System::Clock::Timestamp backoff          = rm->GetBackoff(test->backoffBase, test->sendCount);
```
Comment (andy31415) [inline src/messaging/tests/TestReliableMessageProtocol.cpp:1510]: nit: any value in looping a bit over every test vector to try different jitter values just in case? seems fast enough to compute.

File: src/messaging/ReliableMessageMgr.cpp [hunk 6]
```diff
@@ -187,11 +187,47 @@ CHIP_ERROR ReliableMessageMgr::AddToRetransTable(ReliableMessageContext * rc, Re
...
+    if (retryCount < 0)
+        retryCount = 0; // Enforce floor
+    if (retryCount > 4)
+        retryCount = 4; // Enforce reasonable maximum after 5 tries
+
+    // Implement `t = i⋅MRP_BACKOFF_BASE^max(0,n−MRP_BACKOFF_THRESHOLD)` from Section 4.15.2.1. Retransmissions
```
Comment (andy31415) [inline src/messaging/ReliableMessageMgr.cpp:207]: On my copy the section is 4.11.2.1 instead of 4.15...

File: src/messaging/ReliableMessageMgr.cpp (hunk 4 above)
Comment (turon) [reply src/messaging/ReliableMessageMgr.cpp:203]: I'd rather keep the explicit approach to maximize efficiency given we are dealing with fundamental types and have many embedded system targets.

File: src/messaging/tests/TestReliableMessageProtocol.cpp (hunk 5 above)
Comment (turon) [reply src/messaging/tests/TestReliableMessageProtocol.cpp]: Are you suggesting to feed the test undefined values as a sort of fuzzing technique?  Two issues with that:
1) That wouldn't work because the test vector includes expected values or "truth data", which random values would not get right.
2) Undefined values are unlikely to have an evenly random spread.  A high percentage of the inputs would be simply `0`, as such, explicit fuzz tests are better in those cases.

**Review by andy31415 — APPROVED** [review]

File: src/messaging/tests/TestReliableMessageProtocol.cpp (hunk 5 above)
Comment (andy31415) [reply src/messaging/tests/TestReliableMessageProtocol.cpp]: I was thinking to do each of the test vector test a few times because of the jitter using Rnd data, so to validate we get within min and max every time instead of only once.

Comment (turon) [reply src/messaging/tests/TestReliableMessageProtocol.cpp]: Ah ok.  I can make that change, sure.

File: src/messaging/tests/TestReliableMessageProtocol.cpp [hunk 7]
```diff
@@ -1443,6 +1443,77 @@ void CheckLostStandaloneAck(nlTestSuite * inSuite, void * inContext)
...
+        .backoffMax  = System::Clock::Timestamp(2458),
+    },
+};
+
+const unsigned theBackoffComplianceTestVectorLength =
+    sizeof(theBackoffComplianceTestVector) / sizeof(struct BackoffComplianceTestVector);
+
+void CheckGetBackoff(nlTestSuite * inSuite, void * inContext)
+{
+    for (uint32_t j = 0; j < 3; j++) {
```
Comment (turon) [inline src/messaging/tests/TestReliableMessageProtocol.cpp:1504]:
```suggestion
    // Run 3x iterations to thoroughly test random jitter always results in backoff within bounds.
    for (uint32_t j = 0; j < 3; j++) {
```

**Review by msandstedt — APPROVED** [review]

File: src/messaging/ReliableMessageMgr.cpp [hunk 8]
```diff
@@ -187,11 +187,49 @@ CHIP_ERROR ReliableMessageMgr::AddToRetransTable(ReliableMessageContext * rc, Re
...
+    for (int i = 0; i < retryCount; i++)
+    {
+        backoffNum *= MRP_BACKOFF_BASE_NUMERATOR;
+        backoffDenom *= MRP_BACKOFF_BASE_DENOMENATOR;
+    }
+    backoff = backoff * backoffNum / backoffDenom;
+
+    // Implement jitter scaler: `t *= (1.0+random(0,1)⋅MRP_BACKOFF_JITTER)`
+    // where jitter is random multiplier from 1.000 to 1.250:
+    uint32_t jitter = MRP_BACKOFF_JITTER_BASE + Crypto::GetRandU16() % MRP_BACKOFF_JITTER_MAX;
```
Comment (bzbarsky-apple) [inline src/messaging/ReliableMessageMgr.cpp:221]: This is not entirely uniform in the `[0, MRP_BACKOFF_JITTER_MAX)` range, but close enough I guess....

File: src/messaging/ReliableMessageMgr.h [hunk 9]
```diff
@@ -102,6 +102,16 @@ class ReliableMessageMgr
      */
     CHIP_ERROR AddToRetransTable(ReliableMessageContext * rc, RetransTableEntry ** rEntry);
 
+    /**
+     *  Calculate the backoff timer for the retransmission.
+     *
+     *  @param[in]   backoffBase    The base interval to use for the backoff calculation, either the active or idle interval.
+     *  @param[in]   sendCount      Count of how many times this message has been sent, including the current attempt.
```
Comment (bzbarsky-apple) [inline src/messaging/ReliableMessageMgr.h:109]: If it's "including the current attempt", this counter is always >= 1, so why are we checking for 0 above?

But also, in the spec this seems to be set to 0 for the first transmission as far as I can tell, not 1?  Hard to tell given the spec's phrasing.

And in our impl the way this will work as far as I can tell is that `entry->sendCount` will be 0 after we send the original message and before we send the first retransmit, and then increase after we send each retransmit.

But also, right now this is only being called when computing the delay between original transmission and first retransmit, because that's what `StartRetransmision` does.  `ReliableMessageMgr::ExecuteActions` is where the inter-retransmit delays are computed.

So these comments don't match the implementation (and don't match the spec, I think), and the implementation does not match the spec.

File: src/messaging/ReliableMessageMgr.cpp [hunk 10]
```diff
@@ -187,11 +187,49 @@ CHIP_ERROR ReliableMessageMgr::AddToRetransTable(ReliableMessageContext * rc, Re
...
+}
+
 void ReliableMessageMgr::StartRetransmision(RetransTableEntry * entry)
 {
-    // TODO: Choose active/idle timeout corresponding to the activity of exchanges of the session.
-    entry->nextRetransTime =
-        System::SystemClock().GetMonotonicTimestamp() + entry->ec->GetSessionHandle()->GetMRPConfig().mIdleRetransTimeout;
+    // TODO(#15800): Choose active/idle timeout corresponding to the ActiveState of peer in session.
+    System::Clock::Timestamp backoff =
+        ReliableMessageMgr::GetBackoff(entry->ec->GetSessionHandle()->GetMRPConfig().mActiveRetransTimeout, entry->sendCount);
```
Comment (bzbarsky-apple) [inline src/messaging/ReliableMessageMgr.cpp:231]: Why the change from idle to active here?

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: src/messaging/ReliableMessageMgr.cpp (hunk 8 above)
Comment (turon) [reply src/messaging/ReliableMessageMgr.cpp:221]: Do you have a suggestion for a technique that is embedded efficient that results in even spread?

File: src/messaging/ReliableMessageMgr.h (hunk 9 above)
Comment (turon) [reply src/messaging/ReliableMessageMgr.h:109]: Yes, the spec starts with `0`. I'll revise this comment to make clear that count starts with `0` and computes the delay between the last sent message and the next retransmission.

The check for `0` above is because the backoff is linear for the first MRP_BACKOFF_THRESHOLD attempts (2 given start-from-zero counting).

It sounds like the implementation is close to what we want.  We want:

- 1st transmission to be sent without delay, 
- 2nd after backoff(0), 
- 3rd after backoff(1) [note this will be linear equal to backoff(1) due to threshold above], 
- 4th after backoff(2) [exponential increase will start on this one], and so on.

We certainly want spec, implementation, and comments to all match, so thanks very much for pointing out discrepancies.

File: src/messaging/ReliableMessageMgr.cpp (hunk 10 above)
Comment (msandstedt) [reply src/messaging/ReliableMessageMgr.cpp:231]: I don't think this is conformant to proper sleepy or powered device behavior.  Why would we declare our connection is active if we've never heard back from the peer?

File: src/messaging/ReliableMessageMgr.h (hunk 9 above)
Comment (bzbarsky-apple) [reply src/messaging/ReliableMessageMgr.h:109]: > It sounds like the implementation is close to what we want. We want:

Yes, the impl is doing that, modulo the issue around `ExecuteActions`.

File: src/messaging/ReliableMessageMgr.cpp (hunk 10 above)
Comment (bzbarsky-apple) [reply src/messaging/ReliableMessageMgr.cpp:231]: The current setup (before this change) is that we use idle for the first retransmit and active for subsequent retransmits, fwiw.  I think it's worth keeping that in place in this PR and making changes to that behavior in a separate PR, just as risk-mitigation.

Comment (msandstedt) [reply src/messaging/ReliableMessageMgr.cpp:231]: Yeah, I know and it baffles me that we currently do this.  It is the opposite of back-off.  But avoiding scope creep is fine.

File: src/messaging/ReliableMessageMgr.h (hunk 9 above)
Comment (turon) [reply src/messaging/ReliableMessageMgr.h]: Updated the comment.

File: src/messaging/ReliableMessageMgr.h [hunk 11]
```diff
@@ -102,6 +102,17 @@ class ReliableMessageMgr
      */
     CHIP_ERROR AddToRetransTable(ReliableMessageContext * rc, RetransTableEntry ** rEntry);
 
+    /**
+     *  Calculate the backoff timer for the retransmission.
+     *
+     *  @param[in]   backoffBase    The base interval to use for the backoff calculation, either the active or idle interval.
+     *  @param[in]   sendCount      Count of how many times this message has been sent, including the current retransmission attempt
+     * starting from `0`.
```
Comment (bzbarsky-apple) [inline src/messaging/ReliableMessageMgr.h:110]: How about:
```suggestion
     *  @param[in]   sendCount      Count of how many times this message
     *                              has been retransmitted so far (0 if it has
     *                              been sent only once with no retransmits,
     *                              1 if it has been sent twice, etc).
```

File: src/messaging/ReliableMessageMgr.h (hunk 9 above)
Comment (msandstedt) [reply src/messaging/ReliableMessageMgr.h]: Are changes to `ExecuteActions` planned in this PR, or are we considering that separate work?

Comment (turon) [reply src/messaging/ReliableMessageMgr.h]: Added backoff calculation to ExecuteActions.  Sorry for the confusion.

File: src/messaging/ReliableMessageMgr.cpp (hunk 10 above)
Comment (turon) [reply src/messaging/ReliableMessageMgr.cpp]: Updated `ExecuteAction` and kept use of active there.  #15800 is tracking fixing idle/active tracking of peers by session.

**Review by bzbarsky-apple — APPROVED** [review]
